### DISCUSSION TITLE
Question about expressions

### DISCUSSION THREAD

**Discussion by ghost** [elsa-workflows/elsa-core#1170] [Q&A] [answered]
I have a question about the expressions. I'm noticing that if you define a object parameter on the activity side =>
![image](https://user-images.githubusercontent.com/8503234/123118122-a53b7780-d442-11eb-990f-551a3795e312.png)

Then add this through the desginer =>
![image](https://user-images.githubusercontent.com/8503234/123118305-cdc37180-d442-11eb-89b2-c6422786c71a.png)

I'm always getting a "string" object. Is there a way to define custom types to the variables? Also array types?

**Comment by sfmskywalker** [comment, marked as answer]
This is by design: the Liquid parser **always** returns a string value. It works similar to razor files where even though you can execute C# code, the result is a string value.

Elsa then takes this string value and tries to convert it into a destination type. To determine the destination type, Elsa looks at the property type, which in your case is `object`. This means nothing to do - a string is assignable to object.

If you changed your `Parameter1` property to be e.g. `int`, then Elsa would convert the string `"10"` to an int value of `10`.

This is different for JavaScript expressions - in that case, the `Paramter1` property will actually receive a long or double value (a JS `number` translates to a C# `long` or something similar).

Using JS expressions, you might be able to return a JSON object that includes C# type information which is picked up by the JSON deserializer. For instance, if your JS expression returns:

```json
{
   "$type": "MyProjectNamespace.MyType, MyProjectAssembly",
   "SomeProperty1": 10,
   "SomeProperty2": "Hello"
}
```

Then it's possible for Elsa to deserialize that JSON into an actual `MyType` instance. 

I haven't tried this, so it's possible it doesn't actually work right now but it should be straightforward for me to enable if you need it.

I don't see an elegant way to achieve something similar for Liquid, since there's no way to associate the destination type with the expression.

One thing you could do is add a secondary property to your activity that enables the user to specify the destination type separately, e.g. 

```csharp
public Type Parameter1Type { get; set; }
```

Right now Elsa would render that as a free-form text input control, into which the user can specify the destination type.

Your activity will then have to take care of converting the Parameter1 value into the proper type (possible using the ConvertTo extension method Elsa uses internally).
